src/inference/Inference.jl

```diff
@@ -573,92 +596,188 @@ function assume(
+end
+function _dot_tilde(sampler, right::NamedDist, left::AbstractArray, vn::VarName, vi)
+    name = right.name
+    if name isa String
+        sym_str, inds = split_var_str(name, String)
+        sym = Symbol(sym_str)
+        vn = VarName{sym}(inds)
+    elseif name isa Symbol
+        vn = VarName{name}("")
+    elseif name isa VarName
+        vn = name
     else
-        rs = isa(spl, SampleFromUniform) ? init(dist, n) : rand(dist, n)
+        throw("Unsupported variable name. Please use either a string, symbol or VarName.")
+    end
+    return _dot_tilde(sampler, right.dist, left, vn, vi)
+end
 
-        if isa(dist, UnivariateDistribution) || isa(dist, MatrixDistribution)
-            for i = 1:n
-                push!(vi, vns[i], rs[i], dist, spl)
-            end
-            @assert size(var) == size(rs) "Turing.assume: variable and random number dimension unmatched"
-            var = rs
-        elseif isa(dist, MultivariateDistribution)
-            for i = 1:n
-                push!(vi, vns[i], rs[:,i], dist, spl)
-            end
-            if isa(var, Vector)
-                @assert length(var) == size(rs)[2] "Turing.assume: variable and random number dimension unmatched"
-                for i = 1:n
-                    var[i] = rs[:,i]
-                end
-            elseif isa(var, Matrix)
-                @assert size(var) == size(rs) "Turing.assume: variable and random number dimension unmatched"
-                var = rs
-            else
-                @error("Turing.assume: unsupported variable container"); error()
-            end
+function dot_assume(
+    spl::Union{SampleFromPrior, SampleFromUniform},
+    dist::MultivariateDistribution,
+    vn::VarName,
+    var::AbstractMatrix,
+    vi::VarInfo
+)
+    @assert dim(dist) == size(var, 1)
+    getvn = i -> VarName(vn, vn.indexing * "[:,$i]")
+    vns = getvn.(1:size(var, 2))
+    r = get_and_set_val!(vi, vns, dist, spl)
+    lp = sum(logpdf_with_trans(dist, r, istrans(vi, vns[1])))
+    var .= r
+    return var, lp
+end
+function dot_assume(
+    spl::Union{SampleFromPrior, SampleFromUniform},
+    dists::Union{Distribution, AbstractArray{<:Distribution}},
+    vn::VarName,
+    var::AbstractArray,
+    vi::VarInfo
+)
+    getvn = ind -> VarName(vn, vn.indexing * "[" * join(Tuple(ind), ",") * "]")
+    vns = getvn.(CartesianIndices(var))
+    r = get_and_set_val!(vi, vns, dists, spl)
+    lp = sum(logpdf_with_trans.(dists, r, istrans(vi, vns[1])))
+    var .= r
+    return var, lp
+end
+function dot_assume(
+    spl::Sampler,
+    ::Any,
+    ::VarName,
+    ::Any,
+    ::VarInfo
+)
+    error("[Turing] $(alg_str(spl)) doesn't support vectorizing assume statement")
+end
+
+@inline function get_and_set_val!(vi, vn::VarName, dist::Distribution, spl)
```

Comment:
Is it crucial to inline this?

Reply:
Probably not. I will do another inline vs noinline PR at some point for every function in Turing! Will remove this for now.